### ISSUE TITLE
This repo seems dead. Alternatives?

### ISSUE THREAD

**Issue by chairbender** [keras-rl/keras-rl#164]
No fault on the owner - they created a very useful package which seems to have no better alternative. But I'd like to see development continue on this and would like to contribute if I can.

I see that pull requests have been left unmerged for 2 months and owner seems to not be actively answering questions. At this point we have to consider moving elsewhere if we want to see this continue. Anyone who actually wants to contribute to this will need to find a different way to do so.

Is there already a good candidate fork for moving to which has someone who will actively maintain it?

If not, perhaps I can go through the trouble of tracking down the forks and getting the PRs and other changes merged into my own fork. Actions speak louder than words, of course. But I'd like to opt for someone else's fork if they feel they have a strong grasp on this codebase / area.

**Comment by pollini** [comment]
+1. I'm still using this package/framework for my research because I've found it very easy to begin with. But now that I'm deeper in the subject of RL, I find it very difficult to get the information I need to shape the agent(s) according to my needs (i.e. my custom environments). So imo, documentation is a big point that should be addressed.

Right now, I'm thinking about changing to the tensorforce library which seems to integrate equally well with gym, and has a more active community and maintainers.

**Comment by chairbender** [comment]
@pollini wow, I wish I had heard of that repo sooner. Looks like exactly what I'm looking for.

**Comment by kirkscheper** [comment]
@matthiasplappert I have to agree with what has been said so far. This is a great project but it has limited potential as is. What are your plans for this repo? You seem to have limited time to develop the code and review pull requests. Consider moving this repo to an organisation where you can take on some collaborators to help you review and develop code.

**Comment by lemonteaa** [comment]
@chairbender (and also @kirkscheper ) if you would like to contribute, just fork and work on it, then submit a PR when you're done. The owner started this as a tool for his PhD thesis (if I understand correctly), but he has moved on and can only merge PR that are mostly ready (I think?). Elsewhere (at #132 ) he does plan on handing this project over to the community-at-large so it can continue to be maintained and developed.

@pollini yes, documentation is a pain-point. Right now your best bet if you still depends on this project is to drill down and read the source code through (the codebase's not large) - I've learnt a thing or two from doing that myself. From a purely user's point of view other libraries may indeed fit your need better, so I think what I can do here is to mention that different libraries/framework has different assumptions/contexts.

As I see it, tensorforce is focused on enhancing the appeal of tensorflow (against its competitors such as Microsoft's [CNTK](https://github.com/Microsoft/CNTK)) by providing a tightly integrated experience. If you don't mind being tied to one specific backend (e.g. you're just a researcher who need to get this done yesterday) then that's a perfectly fine choice.
On the other hand this libraries (or should I say framework) is meant to integrate with Keras, which is itself an effort to make neural network backend-agnostic (but tied to Python AFAIK).

So is this project dead? I tend to believe that it is just "slow" relative to what is expected in this field, which has seen explosive growth over the last year. I, for one, haven't given up on my own PR yet (PPO+A2C) - I'm tied down by other circumstances/responsibilities in life recently though... I'll try to figure out what I can do to get my own PR to completion, which is already in the clean-up phase. ;)

**Comment by M00NSH0T** [comment]
I was about to give up on this myself after first getting into RL about 5-6 months back (watching the Silver lectures and reading the Sutton book).  At first my reservation was the lack of documentation and my application was not gaming, so the existing examples were of little help. So I spent a good while looking around for alternative solutions, including tensorforce, but honestly Keras-RL is so much easier to use than anything else I've found, so a few days ago, I came back and within hours, I was able to rig it up to my new custom OpenAI environment. 

If you just look at the examples that tensorforce gives, there's so much more configuration you need to do... they all have this long main function with hundreds of lines in it, vs Keras-RL, where you simply setup a normal Keras neural network, create an environment, and then plug and play with whatever agent or policy you want. I think I'll eventually spend the time to learn the tensorforce way of doing it, but I just don't have time at the moment. With Keras-RL, the main thing you need to do is build your own custom openAI environment, which is challenging given the lack of documentation, but not impossible.  Took me about 2 months to do this for my own project... could have done it in a week or two if documentation had existed, but that's on the OpenAI team to do.

The beauty of Keras-RL is that it's just a matter of changing a couple lines of code to go from using Q learning to SARSA.  I sincerely hope we can drum up some support to keep this thing alive. Hopefully somebody looking to build a resume will take on the management of this. Nothing else I've found is nearly as intuitive to use.  

If we can get this forked and find somebody to act as primary custodian, I have a short wish list:

- I wish there were a few more agents to use.  Like actor / critic for discrete spaces, some policy gradient options, etc. Maybe even some of the basic ones, like TD-lambda and MC. I realize they may not perform as well, but I think they'd be valuable benchmarks at least, and perhaps help those coming from the same Silver / Sutton route get acclimated quicker.
- I also wish there was support for (or at least instructions for using) multi-discrete and tuple action spaces, which are much more intuitive to setup.  I had to create a clunky single discrete space with 2600+ available actions for my environment, and just comment in the code to remind myself what they all do. I'm sure there's a way to use multi-discrete, but I couldn't figure it out in any reasonable amount of time, so calculating out all the combinations and mapping them to a single discrete action space was what I eventually had to do.

**Comment by matthiasplappert** [comment]
As already mentioned, I can't really justify working on this project since I don't use it for my research anymore. I'm glad that people find it useful and would like to keep things going.

I'm pretty bad with responding to PRs and issues, so if someone here is actively using Keras-RL and would like to become an active maintainer, please do reach out to me at matthiasplappert@me.com.

**Comment by flavianh** [comment]
@matthiasplappert Sent you an email about that a few days ago :)

**Comment by RaphaelMeudec** [comment]
Hi everyone! More maintainers have joined to help Matthias. I've noted all your comments and will take it into account for the next iterations of this repo!

**Comment by guillem74** [comment]
@M00NSH0T Do we have any update for this part?

train_on_batch: I think this is what it's doing currently, but I'm not 100%. I'm concerned that if it's not using the Keras train_on_batch function for all updates to the neural network, then if you spend say a week training something with tons of historic data, then get a small amount of additional data and want to update the network, simply loading the neural net weights from the save file, training with the latest dataset, then saving the updated weights may weigh the new set much more than the far greater cache of original data. From my reading, it sounds like the keras train_on_batch function is better for this use case. It would be nice to build that in, if it isn't used already.

@RaphaelMeudec @flavianh @matthiasplappert
